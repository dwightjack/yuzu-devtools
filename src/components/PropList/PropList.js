import { wire } from 'hyperhtml';
import Prop from '../Prop/Prop';
import PropWatcher from '../PropWatcher/PropWatcher';
import { noop } from '../utils';
import * as styles from './PropList.styles';

const empty = (keys) => wire(keys)`<p class="${styles.empty}">empty object</p>`;

export default function PropList({
  uid,
  title = '',
  onSelect = noop,
  props = {},
  watchers = [],
  watchable = false,
}) {
  const keys = Object.keys(props);
  const list =
    keys.length > 0
      ? keys.map((key) => {
          return Prop({
            key,
            uid,
            watchable,
            value: props[key],
            onSelect,
            watched: watchable && watchers.includes(`${uid}:${key}`),
          });
        })
      : empty(keys);

  const globalWatcher =
    keys.length && watchable
      ? PropWatcher({
          watched: watchers.includes(`${uid}:*`),
          uid,
          onSelect,
          key: '*',
        })
      : '';

  return wire(props)`
    <section class="${styles.root}">
      <h3 class="${styles.title}"><span>${title}</span>${globalWatcher}</h3>
      ${list}
    </section>
  `;
}
