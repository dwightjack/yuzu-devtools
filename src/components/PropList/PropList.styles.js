import { css } from 'emotion';

export const root = css`
  & + & {
    margin-top: var(--gutter);
    border-top: 1px solid var(--color-light);
  }
`;

export const title = css`
  margin: var(--gutter);
  font-size: var(--font-size-m);
  font-weight: bold;
  line-height: 1.2;
`;

export const empty = css`
  margin: 0;
  padding: 0 var(--gutter);
  color: var(--color-quiet);
  font-size: var(--font-size-m);
  font-style: italic;
`;
