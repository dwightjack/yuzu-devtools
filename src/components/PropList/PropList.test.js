// import Prop from '../Prop/Prop';
// import PropWatcher from '../PropWatcher/PropWatcher';
import PropList from './PropList';

jest.mock('../Prop/Prop', () => {
  const { mockComponent } = require('../test-utils');
  return jest.fn(mockComponent('Prop'));
});
jest.mock('../PropWatcher/PropWatcher', () => {
  const { mockComponent } = require('../test-utils');
  return jest.fn(mockComponent('PropWatcher'));
});

describe('PropList', () => {
  const props = {
    name: 'John',
    surname: 'Doe',
  };

  test('matches default snapshot', () => {
    expect(
      PropList({
        uid: '_0',
        title: 'TITLE',
        props,
      }),
    ).toMatchSnapshot();
  });

  test('matches watchable snapshot', () => {
    expect(
      PropList({
        uid: '_0',
        title: 'WATCHABLE LIST',
        props,
        watchable: true,
        watchers: ['_0:name'],
      }),
    ).toMatchSnapshot();
  });
});
