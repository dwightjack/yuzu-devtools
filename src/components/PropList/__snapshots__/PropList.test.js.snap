// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropList matches default snapshot 1`] = `
<section
  class="css-1ak8wuw"
  style=""
>
  
      
  <h3
    class="css-9omne8"
    style=""
  >
    <span
      style=""
    >
      TITLE
      
    </span>
    
    
  </h3>
  
      
  <prop
    key="name"
    onSelect={[Function]}
    style=""
    uid="_0"
    value="John"
    watchable={false}
    watched={false}
  />
  <prop
    key="surname"
    onSelect={[Function]}
    style=""
    uid="_0"
    value="Doe"
    watchable={false}
    watched={false}
  />
  
  
    
</section>
`;

exports[`PropList matches watchable snapshot 1`] = `
<section
  class="css-1ak8wuw"
  style=""
>
  
      
  <h3
    class="css-9omne8"
    style=""
  >
    <span
      style=""
    >
      WATCHABLE LIST
      
    </span>
    <propwatcher
      key="*"
      onSelect={[Function]}
      style=""
      uid="_0"
      watched={false}
    />
    
  </h3>
  
      
  <prop
    key="name"
    onSelect={[Function]}
    style=""
    uid="_0"
    value="John"
    watchable={true}
    watched={true}
  />
  <prop
    key="surname"
    onSelect={[Function]}
    style=""
    uid="_0"
    value="Doe"
    watchable={true}
    watched={false}
  />
  
  
    
</section>
`;
